File: code-coaches/hard/longest-common-substring/LongestCommonSubstring.swift
import Foundation

func getLongestCommonSubstring(_ words: [String]) -> String {
    let shortestWord = words.min(by: { $0.count < $1.count }) ?? ""
    let maxLength = shortestWord.count
    var longestSubstring = ""

    for start in 0..<maxLength {
        for end in (start + 1)...maxLength {
            let startIdx = shortestWord.index(shortestWord.startIndex, offsetBy: start)
            let endIdx = shortestWord.index(shortestWord.startIndex, offsetBy: end)
            let substring = String(shortestWord[startIdx..<endIdx])
            if words.allSatisfy({ $0.contains(substring) }) {
                if substring.count > longestSubstring.count
                   || (substring.count == longestSubstring.count && substring < longestSubstring)
                {
                    longestSubstring = substring
                }
            }
        }
    }

    return longestSubstring
}

func main() {
    guard let input = readLine() else { return }
    let words = input.split(separator: " ").map { String($0) }
    print(getLongestCommonSubstring(words))
}

main()

File: code-coaches/hard/longest-common-substring/longest_common_substring.rb
def get_longest_common_substring(words)
  shortest_word = words.min_by(&:length)
  max_length = shortest_word.length
  longest_substring = ""

  (0...max_length).each do |start|
    (start + 1..max_length).each do |end_pos|
      substring = shortest_word[start...end_pos]
      if words.all? { |word| word.include?(substring) }
        if substring.length > longest_substring.length ||
           (substring.length == longest_substring.length && substring < longest_substring)
          longest_substring = substring
        end
      end
    end
  end

  longest_substring
end

def main
  words = gets.chomp.split
  puts get_longest_common_substring(words)
end

if __FILE__ == $0
  main
end
